sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonArrowInput.scala

```diff
@@ -140,3 +141,53 @@ private[python] trait BasicPythonArrowInput extends PythonArrowInput[Iterator[In
     }
   }
 }
+
+
+private[python] trait BatchedPythonArrowInput extends BasicPythonArrowInput {
+  self: BasePythonRunner[Iterator[InternalRow], _] =>
+  private val arrowMaxRecordsPerBatch = SQLConf.get.arrowMaxRecordsPerBatch
+
+  private val maxBytesPerBatch = SQLConf.get.arrowMaxBytesPerBatch
+
+  // Marker inside the input iterator to indicate the start of the next batch.
+  private var nextBatchStart: Iterator[InternalRow] = Iterator.empty
+
+  override protected def writeNextBatchToArrowStream(
+      root: VectorSchemaRoot,
+      writer: ArrowStreamWriter,
+      dataOut: DataOutputStream,
+      inputIterator: Iterator[Iterator[InternalRow]]): Boolean = {
+    if (!nextBatchStart.hasNext) {
+      if (inputIterator.hasNext) {
+        nextBatchStart = inputIterator.next()
+      }
+    }
+    if (nextBatchStart.hasNext) {
+      val startData = dataOut.size()
+
+      var numRowsInBatch: Int = 0
+
+      def underBatchSizeLimit: Boolean =
+        (maxBytesPerBatch == Int.MaxValue) || (arrowWriter.sizeInBytes() < maxBytesPerBatch)
```

Comment:
Due to the API issue https://github.com/apache/spark/pull/50080/files#r1971063958

Maybe we can call `ArrowWriter.bytesWritten` (https://arrow.apache.org/docs/dev/java/reference/org/apache/arrow/vector/ipc/ArrowWriter.html#bytesWritten())?

Reply:
Sounds good. Let me followup for this separately if you don't mind.

Reply:
Okay